agent/consul/state/catalog.go

```diff
@@ -1463,18 +1473,12 @@ func (s *Store) deleteServiceTxn(tx *memdb.Txn, idx uint64, nodeName, serviceID
 				return err
 			}
 
-			// Clean up association between service name and gateway
-			gateway, err := s.serviceTerminatingGateway(tx, svc.ServiceName, &svc.EnterpriseMeta)
-			if err != nil {
-				return fmt.Errorf("failed gateway lookup for %q: %s", svc.ServiceName, err)
+			// Clean up association between service name and gateways
+			if _, err := tx.DeleteAll(gatewayServicesTableName, "service", structs.NewServiceID(svc.ServiceName, entMeta)); err != nil {
```

Comment:
Ah I see what you are saying. Wildcard specifiers really cause all sorts of problems here. 🤔 

I think the option you specify also has a problem because what about the case where a config entry specifies both a wildcard specifier and the actual service name? We should not delete the mapping in that case so that we maintain any different configuration, i.e. specific TLS creds.

Reply:
For our future selves:

We talked about this some more and think a solution may instead be to add a flag to the `GatewayService` that indicates if the record was generated from a wildcard. If it was, we delete it.

Reply:
Would it be possible to have it created via two wildcards and you'd want to retain it when deleting just the one wildcard? In that case you might want some explicit refcounting.